cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java

```diff
@@ -140,4 +157,54 @@ private static URL getURLByClass(ClassLoader classLoader, String className) {
     }
     return ClassLoaders.getClassPathURL(className, resource);
   }
+
+  /**
+   * Starts Py4j gateway server if in executor container and running PySpark.
+   *
+   * @param classLoader the classloader to use for loading classes
+   * @return a {@link Runnable} that calling the {@link Runnable#run()} method will stop the server.
+   */
+  private static Runnable startGatewayServerIfNeeded(ClassLoader classLoader) {
+    Runnable noopRunnable = new Runnable() {
+      @Override
+      public void run() {
+        // no-op
+      }
+    };
+
+    // If we are not running PySpark or
+    // if this process is the AM, no need to start gateway server
+    // The spark execution service uri is always set for the driver (AM) process
+    if (!isPySpark() || System.getenv(SparkRuntimeUtils.CDAP_SPARK_EXECUTION_SERVICE_URI) != null) {
+      return noopRunnable;
+    }
+
+    // Otherwise start the gateway server using reflection. Also write the port number to a local file
+    Path portFile = Paths.get("cdap.py4j.gateway.port.txt");
+    try {
+      final Object server = classLoader.loadClass(SparkPythonUtil.class.getName())
+        .getMethod("startPy4jGateway", Path.class).invoke(null, portFile);
+
+      LOG.info("Py4j GatewayServer started, listening at port {}",
+               new String(Files.readAllBytes(portFile), StandardCharsets.UTF_8));
+
+      return new Runnable() {
+        @Override
+        public void run() {
+          try {
+            server.getClass().getMethod("shutdown").invoke(server);
+          } catch (Exception e) {
+            LOG.warn("Failed to shutdown Py4j GatewayServer", e);
+          }
+        }
+      };
+    } catch (Exception e) {
+      LOG.warn("Failed to start Py4j GatewayServer. No CDAP functionality will be available in executor", e);
```

Comment:
if someone then tries to use CDAP functionality in executor, would it be a clear/useful error message?
Otherwise, I think its better to make this fatal (throw an exception from here).

Reply:
Not sure if we want to fail if the user is using zero CDAP functionalities

Reply:
As long as we make sure the error message is clear to the user if they do try to use CDAP functionalities.